### PR TITLE
Generate CMake load dependencies for gem json dependencies

### LINKED ISSUE #15915: As a gem developer I want a way for O3DE to automatically add my gem dependency targets to RUNTIME_DEPENDENCIES
**Is your feature request related to a problem? Please describe.**
When creating a gem that depends on another gem you need to modify your `gem.json` to include that dependency in the `dependencies` list, but you also need to add `BUILD_DEPENDENCIES` and `RUNTIME_DEPENDENCIES` in your `CMakelists.txt`

If O3DE's cmake provided a function that would automatically added the immediate dependency's targets as `RUNTIME_DEPENDENCIES` it would help developers who forget this step and then wonder why everything is compiling OK, and yet the game or tool crashes because they forgot to add the `RUNTIME_DEPENDENCIES` entry.

**Describe the solution you'd like**
1. a function exists like `o3de_auto_load_dependency_variants(GEM_NAME ${gem_name} VARIANTS Clients Servers Unified Tools Builders)`
2. The above function call is added to DefaultGem template with inline comments documenting what it does and to remove it if the user wants more control over how runtime loading happens.

**Describe alternatives you've considered**
1. better documentation and error checking in cmake or c++
2. better error messages if possible to help developers track down the root cause

**Additional context**
This issue motivated this request: https://github.com/o3de/o3de/issues/15829

### PR DESCRIPTION
**Pull request by lumberyard-employee-dm**
Added `o3de_add_variant_dependencies_for_gem_dependencies` cmake function

The `o3de_add_variant_dependencies_for_gem_dependencies` function will
add CMake TARGET dependencies for a specified Gem and a set of Gem variants.
For each Gem variant specified a check for a TARGET of `${gem_name}.${variant}` will occur on the gem.
If such a TARGET exists, then each of the specified Gem's dependent gems
listed in its gem.json `"dependencies"` field will have the
corresponding `${dependency}.${variant}` TARGET added as a dependency.

For example in the MultiplayerSample project, there is a
[MPSGameLift](https://github.com/o3de/o3de-multiplayersample/blob/development/MPSGameLift/gem.json) gem that has a dependency on the [AWSGameLift](https://github.com/o3de/o3de-multiplayersample/blob/development/MPSGameLift/gem.json#L23) gem.

If the desire for the MPSGameLift gem when active to to have the AWSGameLift gem be
added as a load dependency automatically for the Gem Variants of
`Clients`, `Servers` and `Unified`, then the MPSGameLift gem can use the
`o3de_add_variant_dependencies_for_gem_dependencies` function to add
the following dependencies
```
MPSGameLift.Clients -> AWSGameLift.Clients
MPSGameLift.Servers -> AWSGameLift.Servers
MPSGameLift.Unified -> AWSGameLift.Unified
```

This would cause the `cmake_dependencies.multiplayersample.<launcher>.setreg` files to add both MPSGameLift shared libraries and AWSGameLift shared libraries to the list of gems to load.

fixes #15915 

## How was this PR tested?

Built an SDK layout from an engine where the NvCloth gem uses the `o3de_add_variant_dependencies_for_gem_dependencies` function to add CMake Target dependencies for its gem.json dependencies.

Instantiated a new project using the DefaultProject template and instantiated each Gem Template(Default, Asset, CppTool, Graphics, PythonTool, UnifiedMultiplayer) inside of the new project.
Then the instantiated `DefaultGem` had a dependency of `AssetGem` and `GraphicsGem` added to it in its gem.json "dependencies" field
image

The new project was registered to use the SDK layout and the `DefaultGem` gem was activated for the project.

When the project was configured using CMake, the `cmake_dependencies.<project-name>.*.setreg` files were validated to have the gem variant dependencies for the DefaultGem, AssetGem and GraphicsGem for each Gem variant.
image

### CODE REVIEW

File: cmake/Gems.cmake [hunk 1]
```diff
@@ -57,6 +57,75 @@ function(o3de_find_ancestor_gem_root output_gem_module_root output_gem_name sour
...
+    if(NOT gem_name OR NOT gem_variants)
+        return() # Nothing to do
+    endif()
+
+    # Get the gem path using the gem name as a key
+    get_property(gem_path GLOBAL PROPERTY "@GEMROOT:${gem_name}@")
+    if(NOT gem_path)
+        message(FATAL_ERROR "Unable to locate gem path for Gem \"${gem_name}\"."
+        " Is the gem registered in either the ~/.o3de/o3de_manifest.json, ${LY_ROOT_FOLDER}/engine.json,"
+        " any project.json or any gem.json which itself is registered")
```
Comment (AMZN-alexpete) [inline cmake/Gems.cmake:84]:
```suggestion
        " any project.json or any gem.json which itself is registered?")
```

File: cmake/SettingsRegistry.cmake [hunk 2]
```diff
@@ -69,75 +69,107 @@ function(ly_detect_cycle_through_visitation item visited_items visited_items_var
...
-                list(APPEND all_gem_load_dependencies ${dealias_load_dependency})
-            endif()
-        endforeach()
+    if(manually_added_dependencies)
+        list(APPEND load_dependencies ${manually_added_dependencies})
+        message(VERBOSE "${indent}Gem Target \"${target}\" has direct manually added dependencies of: ${manually_added_dependencies}")
     endif()
 
+    # unset the all_gem_load_dependencies tgo prevent values PARENT_SCOPE
+    # from being used
```
Comment (AMZN-alexpete) [inline cmake/SettingsRegistry.cmake:132]:
```suggestion
    # unset all_gem_load_dependencies to prevent using values from PARENT_SCOPE
```

**Review by AMZN-alexpete — APPROVED** [review]
Nice! Just some super minor comment suggestions

**Comment by AMZN-Gene** [comment]
Does this cause any drastic increases to cmake configure time since there's now a transversal of the gem deps?

**Review by AMZN-Gene — APPROVED** [review]

**Comment by lumberyard-employee-dm** [comment]
> Does this cause any drastic increases to cmake configure time since there's now a transversal of the gem deps?

No, because only the functionality is opt-in, the Gem must use the `o3de_add_variants_for_gem_dependencies` command to look through the dependencies.
Secondly only the direct dependencies in the "dependencies" field of the gem.json are added. 
Recursion does not occur.

This leaves it up to the Gem itself as to whether it wants to add CMake load dependencies for each of it's "dependencies".
The other gems that depend on Gem that opt-in to the functionality cannot affect it's traversal of dependencies.

**Pull request merged**